src/psp2_ui.cpp

```diff
@@ -0,0 +1,190 @@
+	BaseUi() {
+	
+	starttick = sceKernelGetProcessTimeWide() / 1000;
+	frame = 0;
+	zoom_state = 0;
+	trigger_state = false;
+	vita2d_init();
+	current_display_mode.width = width;
+	current_display_mode.height = height;
+	current_display_mode.bpp = 32;
+	const DynamicFormat format(
+		32,
+		0x000000FF,
+		0x0000FF00,
+		0x00FF0000,
+		0xFF000000,
+		PF::Alpha);
+	main_texture = vita2d_create_empty_texture_format(
+												width, height,
+												SCE_GXM_TEXTURE_FORMAT_A8B8G8R8);
+	Bitmap::SetFormat(Bitmap::ChooseFormat(format));
+	main_surface = Bitmap::Create(vita2d_texture_get_datap(main_texture),width, height, vita2d_texture_get_stride(main_texture), format);
+	
+	#ifdef SUPPORT_AUDIO
+		audio_.reset(new Psp2Audio());
+	#endif
+	
+	scePowerSetArmClockFrequency(444);
+	sceCtrlSetSamplingMode(SCE_CTRL_MODE_ANALOG);
+	
+}
+
+Psp2Ui::~Psp2Ui() {
+	vita2d_free_texture(main_texture);
+	vita2d_fini();
+}
+
+void Psp2Ui::Sleep(uint32_t time) {
+	uint64_t usecs = time*1000;
+	sceKernelDelayThread(usecs);
+}
+
+uint32_t Psp2Ui::GetTicks() const {
+	return (sceKernelGetProcessTimeWide() / 1000 - starttick);
+}
+void Psp2Ui::BeginDisplayModeChange() {
+	// no-op
+}
+
+void Psp2Ui::EndDisplayModeChange() {
+	// no-op
+}
+
+void Psp2Ui::Resize(long /*width*/, long /*height*/) {
+	// no-op
+}
+
+void Psp2Ui::ToggleFullscreen() {
+	// no-op
+}
+
+void Psp2Ui::ToggleZoom() {
+	// no-op
+}
+
+bool Psp2Ui::IsFullscreen() {
+	return true;
+}
+
+void Psp2Ui::ProcessEvents() {
+	
+	SceCtrlData input;
+	sceCtrlPeekBufferPositive(0, &input, 1);
+	keys[Input::Keys::Z] = (input.buttons & SCE_CTRL_CROSS);
+	keys[Input::Keys::X] = (input.buttons & SCE_CTRL_CIRCLE);
+	keys[Input::Keys::N8] = (input.buttons & SCE_CTRL_TRIANGLE);
+	keys[Input::Keys::LSHIFT] = (input.buttons & SCE_CTRL_SQUARE);
+	keys[Input::Keys::F12] = (input.buttons & SCE_CTRL_SELECT);
+	keys[Input::Keys::ESCAPE] = (input.buttons & SCE_CTRL_START);
+	keys[Input::Keys::RIGHT] = (input.buttons & SCE_CTRL_RIGHT);
+	keys[Input::Keys::LEFT] = (input.buttons & SCE_CTRL_LEFT);
+	keys[Input::Keys::UP] = (input.buttons & SCE_CTRL_UP);
+	keys[Input::Keys::DOWN] = (input.buttons & SCE_CTRL_DOWN);
+	keys[Input::Keys::F2] = (input.buttons & SCE_CTRL_LTRIGGER);
+	
+	//Resolution changing support
+	bool old_state = trigger_state;
+	trigger_state = (input.buttons & SCE_CTRL_RTRIGGER);
+	if ((trigger_state != old_state) && trigger_state) zoom_state = ((zoom_state + 1) % 3);
+	
+	// Left analog support
+	keys[Input::Keys::JOY_AXIS_X_LEFT] = (input.lx < 50);
+	keys[Input::Keys::JOY_AXIS_X_RIGHT] = (input.lx > 170);
+	keys[Input::Keys::JOY_AXIS_Y_DOWN] = (input.ly > 170);
+	keys[Input::Keys::JOY_AXIS_Y_UP] = (input.ly < 50);
+	
+	// Right analog support for extra buttons
+	if (input.ry > 170) keys[Input::Keys::N1] = true;
+	else if (input.ry < 50) keys[Input::Keys::N3] = true;
+	else if (input.rx > 170) keys[Input::Keys::N5] = true;
+	else if (input.rx < 50) keys[Input::Keys::N9] = true;
```

Comment:
A spanish user on twitter reported that the keys of the right joystick get stuck in "on" state when used.
Looks like he is rìght :)